### ISSUE TITLE
[STORM-1106] Netty Client Connection Attempts should not be limited

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4944]
<p>The workers should not give-up making connection with other workers. This could cause the worker to be blocked forever.</p>

---
<details><summary><i>Originally reported by <a title="Kishor Patil" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kishorvpatil">kishorvpatil</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1106" target="_blank">Netty Client Connection Attempts should not be limited</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Kishor Patil" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kishorvpatil">kishorvpatil</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Blocker
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-10-15T15:48:26+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user kishorvpatil opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1106" title="Netty Client Connection Attempts should not be limited" class="issue-link" data-issue-key="STORM-1106"><del>STORM-1106</del></a> Netty should not limit attempts to reconnect</p>



<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/kishorvpatil/incubator-storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kishorvpatil/incubator-storm</a> storm1106</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/795.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #795</p>

<hr />
<p>commit 7b4f4d7b71e72c5e5be24de8e70e893ebb741ec2<br/>
Author: Kishor Patil <kpatil@yahoo-inc.com><br/>
Date:   2015-10-12T16:31:24Z</p>

<p>    Netty should not limit attempts to reconnect</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-147456715" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-147456715</a></p>

<p>    The change itself looks OK, but I would like us to remove the maxConnectionAttemtps member variable.  It is not needed anymore, and it scares me that we will regress again with it there.  Additionally we should look at updating the config.</p>

<p>    ```<br/>
/**</p>
<ul>
	<li>Netty based messaging: The max # of retries that a peer will perform when a remote is not accessible<br/>
 */<br/>
public static final String STORM_MESSAGING_NETTY_MAX_RETRIES = "storm.messaging.netty.max_retries";<br/>
public static final Object STORM_MESSAGING_NETTY_MAX_RETRIES_SCHEMA = ConfigValidation.IntegerValidator;<br/>
    ```</li>
</ul>


<p>    Describes what the config was doing, but that is fundamentally flawed.  It would be nice to deprecate this and file a follow on JIRA so that we can use a better named/documented config in the backoff calculation</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user kishorvpatil commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-147459755" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-147459755</a></p>

<ul class="alternate" type="square">
	<li>Updated code to mark the configuration as deprecated</li>
	<li>Remove private variable.</li>
	<li>Created follow-up jira <a href="https://issues.apache.org/jira/browse/STORM-1107" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-1107</a></li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-147460971" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-147460971</a></p>

<p>    +1 pending travis</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user kishorvpatil commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-147739297" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-147739297</a></p>

<p>    The Travis build failure on 1.8 seems unrelated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#discussion_r41902116" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#discussion_r41902116</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/messaging/netty/Client.java &#8212;<br/>
    @@ -134,7 +128,7 @@<br/>
     LOG.info("creating Netty Client, connecting to {}:{}, bufferSize: {}", host, port, bufferSize);<br/>
     int messageBatchSize = Utils.getInt(stormConf.get(Config.STORM_NETTY_MESSAGE_BATCH_SIZE), 262144);</p>

<ul class="alternate" type="square">
	<li>maxReconnectionAttempts = Utils.getInt(stormConf.get(Config.STORM_MESSAGING_NETTY_MAX_RETRIES));<br/>
    +int maxReconnectionAttempts = Utils.getInt(stormConf.get(Config.STORM_MESSAGING_NETTY_MAX_RETRIES));
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    If this config does not make sense anymore, should we stop using it in the code?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#discussion_r41903879" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#discussion_r41903879</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/messaging/netty/Client.java &#8212;<br/>
    @@ -134,7 +128,7 @@<br/>
     LOG.info("creating Netty Client, connecting to {}:{}, bufferSize: {}", host, port, bufferSize);<br/>
     int messageBatchSize = Utils.getInt(stormConf.get(Config.STORM_NETTY_MESSAGE_BATCH_SIZE), 262144);</p>

<ul class="alternate" type="square">
	<li>maxReconnectionAttempts = Utils.getInt(stormConf.get(Config.STORM_MESSAGING_NETTY_MAX_RETRIES));<br/>
    +int maxReconnectionAttempts = Utils.getInt(stormConf.get(Config.STORM_MESSAGING_NETTY_MAX_RETRIES));
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    > It would be nice to deprecate this and file a follow on JIRA so that we can use a better named/documented config in the backoff calculation</p>

<p>    Disregard: I see from the above comment what we intend to do.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-147801288" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-147801288</a></p>

<p>    +1 I would be good to address <a href="https://issues.apache.org/jira/browse/STORM-1107" title="Remove deprecated Config STORM_MESSAGING_NETTY_MAX_RETRIES and fix Netty Client backoff calculations" class="issue-link" data-issue-key="STORM-1107"><del>STORM-1107</del></a>(<a href="https://issues.apache.org/jira/browse/STORM-1107" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-1107</a>) before doing a subsequent release.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-147896620" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-147896620</a></p>

<p>    @kishorvpatil <br/>
    For confirming that I understood this issue well, I'd like you to elaborate this issue some more.</p>

<p>    As far as I understand, when retry count exceeds, Connect.run() throws RuntimeException, but worker doesn't be killed since it is a TimerTask.<br/>
    So it just closes the connection and wait for reassign for such worker.</p>

<p>    If Nimbus reassigns dead worker to another after retry limit exceed, another connection is being made and it would be fine.<br/>
    But some reason if problematic worker is just not able to connect (for example, STW, and so on) to another workers for longer than connection retrying but not forever, and nimbus doesn't reassign problematic worker, another workers cannot connect to problematic worker forever.</p>

<p>    Is my assumption right? Or there's other reason?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user kishorvpatil commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-148208528" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-148208528</a></p>

<p>    1. Nimbus would not reassign anything because - workers come-up and heartbeat correctly ( as it is independent thread.)<br/>
    2. If workerA runs out of attempts to WorkerB ( because workerB was delayed in launching..) The workerA simply hangs..)</p>

<p>    This patch simply helps resolve that bottleneck for trying to connect forever.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user kishorvpatil commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-148209299" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-148209299</a></p>

<p>    Also, there is bigger problem with `HashedWheelTimer` has bigger problems when worker has large number of connections to make and the remote workers are not accepting connections. At least for version of netty we are using there have been reports of it cracking over high CPU usage - i.e. once I saw `HashedWheelTimer` cause worker hangups. I am in the process of creating a reproducible test and solution for that. But we don't have to hold back 0.10.x release for that.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-148210871" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-148210871</a></p>

<p>    @kishorvpatil <br/>
    Thanks for the detail. I didn't think about situation 2. <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-148217847" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-148217847</a></p>

<p>    Seems like default value of storm.messaging.netty.max_wait_ms is very small.</p>

<p>    I didn't see the details about StormBoundedExponentialBackoffRetry, but if it is running as intended via class comment, we're sleeping between 100 ~ 1000 ms for each retry by default.<br/>
    But we're pinning max retry count to 29 from ExponentialBackoffRetry.validateMaxRetries() since we use 1 << retryCount to sleep, resulting that we only sleep about max 29 secs, which is too short.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user kishorvpatil commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-148219849" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-148219849</a></p>

<p>    @HeartSaVioR, your understanding about `StormBoundedExponentialBackoffRetry` is correct, the 29 seconds is decent time - in case where workerB restarts due to any reason - during regular set-up.  But initial startup ( requires supervisor to read assignments/download code/launch workers.) which could mean much more than that. </p>

<p>    The `StormBoundedExponentialBackoffRetry` lets you go beyond max attempts to give you 1sec sleep until you succeed. So, the workerA should keep trying unless nimbus informs workerA that workerB ( is no longer working on that port.)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/795#issuecomment-148222144" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795#issuecomment-148222144</a></p>

<p>    @kishorvpatil <br/>
    Yeah, after this PR gets merged. <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
    Thinking it once more, since default nimbus.task.timeout.secs is 30, retrying for each 1 secs seems not be short.<br/>
    Just curious about adjusting max_wait_ms may affect HashedWheelTimer's CPU usage as you stated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/795" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/795</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kishorvpatil">kishorvpatil</a>:</i>
<p>This was merged into master and 0.10.x-branch.</p>
